### ISSUE TITLE
Incorrect disambiguation by (short) title using MHRA style

### ISSUE THREAD

**Issue by njbart** [jgm/pandoc-citeproc#63]
Disambiguation by (short) title using `modern-humanities-research-association.csl` is not as expected:

Citing like in the following line, disambiguation in subsequent citations is as expected (i.e., last name plus short title):

`Foo [@doe1]. Bar [@doe2]. Foo [@doe1]. Bar [@doe2].`

Citing like in the following, when `@doe2` appears only once, the subsequent `@doe1` is rendered without disambiguation (i.e., just the last name):

`Foo [@doe1]. Bar [@doe2]. Foo [@doe1].`

The latter works as expected using Zotero with `modern-humanities-research-association.csl`. Interestingly enough, pandoc with `chicago-fullnote-bibliography.csl` also gives the expected output.

Example file at https://gist.github.com/njbart/76e01842bc1e5695d72c

**Comment by jgm** [comment]
Here's a test case
```
@item1

@item2

@item1

# References

---
references:

- id: item1
  type: webpage
  author:
  - family: Doe
    given: John
  title: Title
  container-title: The Web Site
  URL: http://www.example.com
  issued:
   date-parts:
    - - 2006
      - 10
      - 26

- id: item2
  type: webpage
  author:
  - family: Doe
    given: Sam
  title: Title
  container-title: The Web Site
  URL: http://www.example.com
  issued:
   date-parts:
    - - 2006
      - 10
      - 26
    - - 2006
      - 11
      - 27
...
```

This gives (plain output):

```
John Doe[1]

Sam Doe[2]

Doe[3]



REFERENCES


Doe, John, ‘Title’, _The Web Site_, 2006 <http://www.example.com>

Doe, Sam, ‘Title’, _The Web Site_, 2006 <http://www.example.com>

[1] ‘Title’, _The Web Site_, 2006 <http://www.example.com>.

[2] ‘Title’, _The Web Site_, 2006 <http://www.example.com>.

[3] 
```
Note also the empty note 3!

**Comment by rriemann** [comment]
Pandoc can be very useful for lawyers and editors of legal texts (e.g. in the public sector). While images and tables are rarely used, citations and footnotes are super-important. Commonly, citations are put into footnotes, for instance in the case of the Oxford oscola CSL style, and ibid. is used extensively.

- https://www.law.ox.ac.uk/research-subject-groups/publications/oscola-styles-endnote-latek-refworks-and-zotero
- http://editor.citationstyles.org/styleInfo/?styleId=http%3A%2F%2Fwww.zotero.org%2Fstyles%2Foscola

**Comment by jgm** [comment]
Current output is a bit different from what was reported above:
```
John Doe[1]

Sam Doe[2]

Doe



REFERENCES


Doe, John, ‘Title’, _The Web Site_, 2006 <http://www.example.com>

Doe, Sam, ‘Title’, _The Web Site_, 2006 <http://www.example.com>

[1] ‘Title’, _The Web Site_, 2006 <http://www.example.com>.

[2] ‘Title’, _The Web Site_, 2006 <http://www.example.com>.
```
Same problem with the ambiguous 'Doe', but at least the empty footnote has been taken care of.

**Comment by cormacrelf** [comment]
As noted over at https://groups.google.com/forum/#!topic/pandoc-discuss/muR8EqxfTik, I'm building a new Rust-based citeproc. I haven't implemented disambiguation yet, but I wrote up how I'm interpreting the spec so far, and it would be good to get some more eyes on it. And maybe some insight into exactly what Disamb.hs is doing, because I can't say I understand it yet.

https://github.com/cormacrelf/citeproc-rs/blob/master/docs/disambiguation.md

**Comment by cormacrelf** [comment]
FYI, I'm getting closer to the implementation described in the previous comment, with the exception of names. This is some output from processing `item1` with MHRA, where the `state.tokens` is the set of disambiguation tokens it encountered while processing, and `matching_ids` are the ones from the library that appeared in the inverted index for those tokens. It seems to work, is all.

```
[citeproc/src/driver.rs:49] state = IrState {
    tokens: {
        Date(
            Date {
                year: 2006,
                month: 10,
                day: 26
            }
        ),
        Date(
            Date {
                year: 2006,
                month: 10,
                day: 0
            }
        ),
        Str(
            Atom('Title' type=inline)
        ),
        Str(
            Atom('http://www.example.com' type=dynamic)
        ),
        Str(
            Atom('The Web Site' type=dynamic)
        ),
        Date(
            Date {
                year: 2006,
                month: 0,
                day: 0
            }
        )
    }
}
[citeproc/src/driver.rs:65] matching_ids = {
    Atom('item1' type=inline),
    Atom('item2' type=inline)
}
```

**Comment by cormacrelf** [comment]
Oh, and I figured out what's going wrong with pandoc-citeproc. If you put the references in an external CSL-JSON file and remove the inline references, it works perfectly. So something's going wrong reading all the MetaInlines, or comparing them.

https://github.com/cormacrelf/citeproc-rs/blob/master/citeproc/tests/data/disamb.json

```
John Doe[1]

Sam Doe[2]

John Doe

REFERENCES


Doe, John, ‘Title’, _The Web Site_, 2006 <http://www.example.com>

Doe, Sam, ‘Title’, _The Web Site_, 2006 <http://www.example.com>

[1] ‘Title’, _The Web Site_, 2006 <http://www.example.com>.

[2] ‘Title’, _The Web Site_, 2006 <http://www.example.com>.

```

**Comment by njbart** [comment]
@cormacrelf Can”t reproduce that putting the references in an external CSL-JSON file would solve this issue. Your input seems to contain `Foo [@item1]. Bar [@item2].` only, not citing `@item1` a second time. Using this script with your `disamb.json`:

```
pandoc -F pandoc-citeproc -t plain << EOT

Foo [@item1].
Bar [@item2].
Foo [@item1].
<!-- Comment out this line to see the problem [@item2]. -->

# References

---
csl: modern-humanities-research-association.csl
bibliography: disamb.json
...
EOT
```

… I still get this output:

```
Foo.[1] Bar.[2] Foo.[3]



REFERENCES


Doe, John, ‘Title’, _The Web Site_, 2006 <http://www.example.com>

Doe, Sam, ‘Title’, _The Web Site_, 2006 <http://www.example.com>

[1] John Doe, ‘Title’, _The Web Site_, 2006 <http://www.example.com>.

[2] Sam Doe, ‘Title’, _The Web Site_, 2006 <http://www.example.com>.

[3] Doe.
```

**Comment by cormacrelf** [comment]
Hm, neither can I. I don't know what happened there, but I am sure I got that result somehow, and piped it to `pbcopy`. And I never added "John " or "John Doe" to the file (it's still open with undo history). It also couldn't have come from citeproc-rs, because it hasn't got AuthorInText support yet.  One of those mysteries. Sorry for waking up the thread again.

**Comment by njbart** [comment]
The issue reported in the original post seems to have been fixed in pandoc 2.11. Closing.
